Data/ByteString.hs

```diff
@@ -1508,8 +1508,44 @@ tails p | null p    = [empty]
 -- ---------------------------------------------------------------------
 -- ** Ordered 'ByteString's
 
--- | /O(n)/ Sort a ByteString efficiently, using counting sort.
+-- | /O(n)/ Sort a ByteString efficiently, using insertion sort for short
+-- 'ByteString's and counting sort for long 'ByteString's.
 sort :: ByteString -> ByteString
+sort bs@(PS _ _ 0) = bs
+
+sort (PS input s l) | l <= threshold = unsafeCreate l $ \p ->
+    withForeignPtr input (\x -> insertionSort p (x `plusPtr` s) l)
+  where
+    -- threshold is where we find insertion sort still outperforms counting
+    -- sort even for worst case.
```

Comment:
It would be nice to clarify, what is the worst case. Could you possibly commit benchmarks, which you used to derive this threshold?